// @directus/cli/dist/cli/commands/whoami.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
const command_1 = require("../../core/command");
exports.default = (0, command_1.command)({
    group: 'utils',
    description: 'Shows connected user information',
    usage: `
			\`\`\`
			$ $0 whoami
			\`\`\`
		`,
    documentation: `
			Shows information about the connected account.
		`,
    features: {
        sdk: true,
        query: 'one',
    },
    hints: ['who', 'me'],
}, async function ({ output, query, sdk }) {
    const user = await sdk.users.me.read({
        fields: ['*', '*.*'],
        ...query.one,
    });
    await output.compose(async (ui) => {
        await ui.wrap(async (ui) => {
            const head = chalk_1.default.green.bold;
            const missing = chalk_1.default.italic.gray('No information');
            await ui.header('User');
            await ui.skip();
            await ui.table([
                [head('Name'), user.first_name ?? missing],
                [head('Email'), user.email ?? missing],
                [head('Role'), user.role?.name ?? missing],
            ]);
        }, 1);
    });
    return user;
});
